File: sources/net.sf.j2s.core/src/net/sf/j2s/core/builder/BuildNotifier.java
/*******************************************************************************
 * Copyright (c) 2000, 2008 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *******************************************************************************/
package net.sf.j2s.core.builder;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Calendar;
import java.util.logging.Formatter;
import java.util.logging.Level;
import java.util.logging.LogRecord;

import org.eclipse.core.resources.*;
import org.eclipse.core.runtime.*;

import org.eclipse.jdt.core.compiler.CategorizedProblem;
import org.eclipse.jdt.core.compiler.IProblem;
import org.eclipse.jdt.internal.compiler.problem.AbortCompilation;
import org.eclipse.jdt.internal.core.util.Messages;
import org.eclipse.swt.SWT;
import org.eclipse.swt.graphics.Color;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.MessageBox;
import org.eclipse.swt.widgets.Shell;
import org.eclipse.ui.PlatformUI;
import org.eclipse.ui.console.ConsolePlugin;
import org.eclipse.ui.console.IConsole;
import org.eclipse.ui.console.MessageConsole;
import org.eclipse.ui.console.MessageConsoleStream;

public class BuildNotifier {

protected IProgressMonitor monitor;
protected boolean cancelling;
protected float percentComplete;
protected float progressPerCompilationUnit;
protected int newErrorCount;
protected int fixedErrorCount;
protected int newWarningCount;
protected int fixedWarningCount;
protected int workDone;
protected int totalWork;
protected String previousSubtask;
private static final String J2S_DEPLOY_MODE_BROWSER = "browser";
private static final String J2S_DEPLOY_MODE_TIZEN = "tizen";
private static final String EXPORT_TAG = "Export";
private String j2sDeployMode;
private IProject currentProject;

private static MessageConsole sConsole;
private static MessageConsoleStream sNormalStream;
private static MessageConsoleStream sErrorStream;
private static Color sColorRed;
private static Formatter sFormatter;

public static int NewErrorCount = 0;
public static int FixedErrorCount = 0;
public static int NewWarningCount = 0;
public static int FixedWarningCount = 0;

public static void resetProblemCounters() {
	NewErrorCount = 0;
	FixedErrorCount = 0;
	NewWarningCount = 0;
	FixedWarningCount = 0;
}

public BuildNotifier(IProgressMonitor monitor, IProject project, String deployMode) {
	this.j2sDeployMode = deployMode;
	this.currentProject = project;
	
	this.monitor = monitor;
	this.cancelling = false;
	this.newErrorCount = NewErrorCount;
	this.fixedErrorCount = FixedErrorCount;
	this.newWarningCount = NewWarningCount;
	this.fixedWarningCount = FixedWarningCount;
	this.workDone = 0;
	this.totalWork = 1000000;
	initConsoleView();
}

/**
 * init MPTConsole for J2S
 */

public void initConsoleView(){
	IConsole[] IConsole = ConsolePlugin.getDefault().getConsoleManager().getConsoles();
	this.sConsole = (MessageConsole) IConsole[1];
	this.sNormalStream = sConsole.newMessageStream();
	this.sErrorStream = sConsole.newMessageStream();
	Display display = ConsolePlugin.getDefault().getWorkbench().getDisplay();
	sColorRed = new Color(display, 0xFF, 0x00, 0x00);
    display.asyncExec(new Runnable() {
        public void run() {
        	sErrorStream.setColor(sColorRed);
        }
    });
	this.sFormatter = new Formatter(){
		public String format(LogRecord record) {
			StringBuffer buffer = new StringBuffer();
			buffer.append(String.format("[%1$tF_%1$tT:%1$tL] [%2$s] [%3$s] %4$s\n", 
                                        Calendar.getInstance(),
                                        record.getLevel(),
                                        record.getLoggerName(),
                                        record.getMessage()));
			if(record.getThrown() != null){
				StringWriter writer = new StringWriter();
				record.getThrown().printStackTrace(new PrintWriter(writer));
				buffer.append(writer.toString());
			}
			return buffer.toString();
		}
	};
}

/**
 * Notification before a compile that a unit is about to be compiled.
 */
public void aboutToCompile(SourceFile unit) {
	String message = Messages.bind(Messages.build_compiling, unit.resource.getFullPath().removeLastSegments(1).makeRelative().toString());
	subTask(message);
}

public void begin() {
	if (this.monitor != null)
		this.monitor.beginTask("", this.totalWork); //$NON-NLS-1$
	this.previousSubtask = null;
}

/**
 * Check whether the build has been canceled.
 */
public void checkCancel() {
	if (this.monitor != null && this.monitor.isCanceled())
		throw new OperationCanceledException();
}

/**
 * Check whether the build has been canceled.
 * Must use this call instead of checkCancel() when within the compiler.
 */
public void checkCancelWithinCompiler() {
	if (this.monitor != null && this.monitor.isCanceled() && !this.cancelling) {
		// Once the compiler has been canceled, don't check again.
		setCancelling(true);
		// Only AbortCompilation can stop the compiler cleanly.
		// We check cancelation again following the call to compile.
		throw new AbortCompilation(true, null);
	}
}

/**
 * Notification while within a compile that a unit has finished being compiled.
 */
public void compiled(SourceFile unit) {
	String message = Messages.bind(Messages.build_compiling, unit.resource.getFullPath().removeLastSegments(1).makeRelative().toString());
	subTask(message);
	updateProgressDelta(this.progressPerCompilationUnit);
	checkCancelWithinCompiler();
}

public void done() {
	NewErrorCount = this.newErrorCount;
	FixedErrorCount = this.fixedErrorCount;
	NewWarningCount = this.newWarningCount;
	FixedWarningCount = this.fixedWarningCount;

	updateProgress(1.0f);
	subTask(Messages.build_done);
	if (this.monitor != null)
		this.monitor.done();
	this.previousSubtask = null;
	
	if (J2S_DEPLOY_MODE_TIZEN.equals(j2sDeployMode)){
		// Message with ok and cancel button and info icon
		if (NewErrorCount != 0) {
			error(EXPORT_TAG, "Project '%1$s' could not be exported for '%2$s'", currentProject.getName(),j2sDeployMode);
			PlatformUI.getWorkbench().getDisplay().asyncExec(new Runnable() {
			    public void run() {
				    Shell activeShell = PlatformUI.getWorkbench().getActiveWorkbenchWindow().getShell();
				    MessageBox dialog = 
				    		  new MessageBox(activeShell, SWT.ICON_QUESTION | SWT.OK | SWT.CANCEL);
				    		dialog.setText("Detect some compilation errors");
				    		dialog.setMessage("Do you really want to ignore these compile errors?");
				    		int returnCode = dialog.open();
				    		if (SWT.OK == returnCode) {
				    			// TODO luqiang, Package for Tizen Project
//				    			performTizenPackage();
				    		}
				}
			});
		}else{
			general(EXPORT_TAG, "Project '%1$s' has been exported for '%2$s' successfully.", currentProject.getName(),j2sDeployMode);
		}
	}else if(J2S_DEPLOY_MODE_BROWSER.equals(j2sDeployMode)){
		if (NewErrorCount != 0) {
			error(EXPORT_TAG, "Project '%1$s' could not be exported for '%2$s'", currentProject.getName(),j2sDeployMode);
		}else{
			general(EXPORT_TAG, "Project '%1$s' has been exported for '%2$s' successfully.", currentProject.getName(),j2sDeployMode);
		}
	}
	
	try {
		currentProject.refreshLocal(IResource.DEPTH_INFINITE, null);
	} catch (CoreException e) {
		// TODO Auto-generated catch block
		e.printStackTrace();
	}
	
}





/**
 * Returns a string describing the problems.
 */
protected String problemsMessage() {
	int numNew = this.newErrorCount + this.newWarningCount;
	int numFixed = this.fixedErrorCount + this.fixedWarningCount;
	if (numNew == 0 && numFixed == 0) return ""; //$NON-NLS-1$

	boolean displayBoth = numNew > 0 && numFixed > 0;
	StringBuffer buffer = new StringBuffer();
	buffer.append('(');
	if (numNew > 0) {
		// (Found x errors + y warnings)
		buffer.append(Messages.build_foundHeader);
		buffer.append(' ');
		if (displayBoth || this.newErrorCount > 0) {
			if (this.newErrorCount == 1)
				buffer.append(Messages.build_oneError);
			else
				buffer.append(Messages.bind(Messages.build_multipleErrors, String.valueOf(this.newErrorCount)));
			if (displayBoth || this.newWarningCount > 0)
				buffer.append(" + "); //$NON-NLS-1$
		}
		if (displayBoth || this.newWarningCount > 0) {
			if (this.newWarningCount == 1)
				buffer.append(Messages.build_oneWarning);
			else
				buffer.append(Messages.bind(Messages.build_multipleWarnings, String.valueOf(this.newWarningCount)));
		}
		if (numFixed > 0)
			buffer.append(", "); //$NON-NLS-1$
	}
	if (numFixed > 0) {
		// (Fixed x errors + y warnings) or (Found x errors + y warnings, Fixed x + y)
		buffer.append(Messages.build_fixedHeader);
		buffer.append(' ');
		if (displayBoth) {
			buffer.append(String.valueOf(this.fixedErrorCount));
			buffer.append(" + "); //$NON-NLS-1$
			buffer.append(String.valueOf(this.fixedWarningCount));
		} else {
			if (this.fixedErrorCount > 0) {
				if (this.fixedErrorCount == 1)
					buffer.append(Messages.build_oneError);
				else
					buffer.append(Messages.bind(Messages.build_multipleErrors, String.valueOf(this.fixedErrorCount)));
				if (this.fixedWarningCount > 0)
					buffer.append(" + "); //$NON-NLS-1$
			}
			if (this.fixedWarningCount > 0) {
				if (this.fixedWarningCount == 1)
					buffer.append(Messages.build_oneWarning);
				else
					buffer.append(Messages.bind(Messages.build_multipleWarnings, String.valueOf(this.fixedWarningCount)));
			}
		}
	}
	buffer.append(')');
	return buffer.toString();
}

/**
 * Sets the cancelling flag, which indicates we are in the middle
 * of being cancelled.  Certain places (those callable indirectly from the compiler)
 * should not check cancel again while this is true, to avoid OperationCanceledException
 * being thrown at an inopportune time.
 */
public void setCancelling(boolean cancelling) {
	this.cancelling = cancelling;
}

/**
 * Sets the amount of progress to report for compiling each compilation unit.
 */
public void setProgressPerCompilationUnit(float progress) {
	this.progressPerCompilationUnit = progress;
}

public void subTask(String message) {
	String pm = problemsMessage();
	String msg = pm.length() == 0 ? message : pm + " " + message; //$NON-NLS-1$

	if (msg.equals(this.previousSubtask)) return; // avoid refreshing with same one
	//if (JavaBuilder.DEBUG) System.out.println(msg);
	if (this.monitor != null)
		this.monitor.subTask(msg);

	this.previousSubtask = msg;
}

protected void updateProblemCounts(CategorizedProblem[] newProblems) {
	for (int i = 0, l = newProblems.length; i < l; i++)
		if (newProblems[i].isError()) this.newErrorCount++; else this.newWarningCount++;
}

/**
 * Update the problem counts from one compilation result given the old and new problems,
 * either of which may be null.
 */
protected void updateProblemCounts(IMarker[] oldProblems, CategorizedProblem[] newProblems) {
	if (newProblems != null) {
		next : for (int i = 0, l = newProblems.length; i < l; i++) {
			CategorizedProblem newProblem = newProblems[i];
			if (newProblem.getID() == IProblem.Task) continue; // skip task
			boolean isError = newProblem.isError();
			String message = newProblem.getMessage();

			if (oldProblems != null) {
				for (int j = 0, m = oldProblems.length; j < m; j++) {
					IMarker pb = oldProblems[j];
					if (pb == null) continue; // already matched up with a new problem
					boolean wasError = IMarker.SEVERITY_ERROR
						== pb.getAttribute(IMarker.SEVERITY, IMarker.SEVERITY_ERROR);
					if (isError == wasError && message.equals(pb.getAttribute(IMarker.MESSAGE, ""))) { //$NON-NLS-1$
						oldProblems[j] = null;
						continue next;
					}
				}
			}
			if (isError) this.newErrorCount++; else this.newWarningCount++;
		}
	}
	if (oldProblems != null) {
		next : for (int i = 0, l = oldProblems.length; i < l; i++) {
			IMarker oldProblem = oldProblems[i];
			if (oldProblem == null) continue next; // already matched up with a new problem
			boolean wasError = IMarker.SEVERITY_ERROR
				== oldProblem.getAttribute(IMarker.SEVERITY, IMarker.SEVERITY_ERROR);
			String message = oldProblem.getAttribute(IMarker.MESSAGE, ""); //$NON-NLS-1$

			if (newProblems != null) {
				for (int j = 0, m = newProblems.length; j < m; j++) {
					CategorizedProblem pb = newProblems[j];
					if (pb.getID() == IProblem.Task) continue; // skip task
					if (wasError == pb.isError() && message.equals(pb.getMessage()))
						continue next;
				}
			}
			if (wasError) this.fixedErrorCount++; else this.fixedWarningCount++;
		}
	}
}

public void updateProgress(float newPercentComplete) {
	if (newPercentComplete > this.percentComplete) {
		this.percentComplete = Math.min(newPercentComplete, 1.0f);
		int work = Math.round(this.percentComplete * this.totalWork);
		if (work > this.workDone) {
			if (this.monitor != null)
				this.monitor.worked(work - this.workDone);
			//if (JavaBuilder.DEBUG)
				//System.out.println(java.text.NumberFormat.getPercentInstance().format(this.percentComplete));
			this.workDone = work;
		}
	}
}

public void updateProgressDelta(float percentWorked) {
	updateProgress(this.percentComplete + percentWorked);
}

/**
 * Print general message to normal console stream
 * @param tag     Component tag
 * @param format  Format string
 * @param args    Format arguments
 */
public void general(String tag, String format, Object ...args){
	LogRecord record = new LogRecord(Level.INFO, String.format(format, args));
	record.setLoggerName(tag);
	sNormalStream.print(sFormatter.format(record));
	ConsolePlugin.getDefault().getConsoleManager().showConsoleView(sConsole);
}

/**
 * Print warning message to error console stream
 * @param tag     Component tag
 * @param format  Format string
 * @param args    Format arguments
 */
public void warning(String tag, String format, Object ...args){
	LogRecord record = new LogRecord(Level.WARNING, String.format(format, args));
	record.setLoggerName(tag);
	sErrorStream.print(sFormatter.format(record));
	ConsolePlugin.getDefault().getConsoleManager().showConsoleView(sConsole);
}

/**
 * Print error message to error console stream
 * @param tag     Component tag
 * @param format  Format string
 * @param args    Format arguments
 */
public void error(String tag, String format, Object ...args){
	LogRecord record = new LogRecord(Level.SEVERE, String.format(format, args));
	record.setLoggerName(tag);
	sErrorStream.print(sFormatter.format(record));
	ConsolePlugin.getDefault().getConsoleManager().showConsoleView(sConsole);
}
}

File: sources/net.sf.j2s.core/bin/net/sf/j2s/core/builder/BuildNotifier.js
Clazz.declarePackage ("net.sf.j2s.core.builder");
Clazz.load (null, "net.sf.j2s.core.builder.BuildNotifier", ["java.lang.StringBuffer", "net.sf.j2s.core.utils.CorePluginConsole", "org.eclipse.core.runtime.OperationCanceledException", "org.eclipse.jdt.internal.compiler.problem.AbortCompilation", "org.eclipse.jdt.internal.core.util.Messages", "$wt.widgets.MessageBox", "org.eclipse.ui.PlatformUI"], function () {
c$ = Clazz.decorateAsClass (function () {
this.monitor = null;
this.cancelling = false;
this.percentComplete = 0;
this.progressPerCompilationUnit = 0;
this.newErrorCount = 0;
this.fixedErrorCount = 0;
this.newWarningCount = 0;
this.fixedWarningCount = 0;
this.workDone = 0;
this.totalWork = 0;
this.previousSubtask = null;
this.j2sDeployMode = null;
this.builderMode = null;
this.currentProject = null;
Clazz.instantialize (this, arguments);
}, net.sf.j2s.core.builder, "BuildNotifier");
c$.resetProblemCounters = Clazz.defineMethod (c$, "resetProblemCounters", 
function () {
($t$ = net.sf.j2s.core.builder.BuildNotifier.NewErrorCount = 0, net.sf.j2s.core.builder.BuildNotifier.prototype.NewErrorCount = net.sf.j2s.core.builder.BuildNotifier.NewErrorCount, $t$);
($t$ = net.sf.j2s.core.builder.BuildNotifier.FixedErrorCount = 0, net.sf.j2s.core.builder.BuildNotifier.prototype.FixedErrorCount = net.sf.j2s.core.builder.BuildNotifier.FixedErrorCount, $t$);
($t$ = net.sf.j2s.core.builder.BuildNotifier.NewWarningCount = 0, net.sf.j2s.core.builder.BuildNotifier.prototype.NewWarningCount = net.sf.j2s.core.builder.BuildNotifier.NewWarningCount, $t$);
($t$ = net.sf.j2s.core.builder.BuildNotifier.FixedWarningCount = 0, net.sf.j2s.core.builder.BuildNotifier.prototype.FixedWarningCount = net.sf.j2s.core.builder.BuildNotifier.FixedWarningCount, $t$);
});
Clazz.makeConstructor (c$, 
function (monitor, project, deployMode, builderMode) {
this.j2sDeployMode = deployMode;
this.currentProject = project;
this.builderMode = builderMode;
this.monitor = monitor;
this.cancelling = false;
this.newErrorCount = net.sf.j2s.core.builder.BuildNotifier.NewErrorCount;
this.fixedErrorCount = net.sf.j2s.core.builder.BuildNotifier.FixedErrorCount;
this.newWarningCount = net.sf.j2s.core.builder.BuildNotifier.NewWarningCount;
this.fixedWarningCount = net.sf.j2s.core.builder.BuildNotifier.FixedWarningCount;
this.workDone = 0;
this.totalWork = 1000000;
}, "org.eclipse.core.runtime.IProgressMonitor,org.eclipse.core.resources.IProject,~S,~S");
Clazz.defineMethod (c$, "aboutToCompile", 
function (unit) {
var message = org.eclipse.jdt.internal.core.util.Messages.bind (org.eclipse.jdt.internal.core.util.Messages.build_compiling, unit.resource.getFullPath ().removeLastSegments (1).makeRelative ().toString ());
this.subTask (message);
}, "net.sf.j2s.core.builder.SourceFile");
Clazz.defineMethod (c$, "begin", 
function () {
if (this.monitor != null) this.monitor.beginTask ("", this.totalWork);
this.previousSubtask = null;
});
Clazz.defineMethod (c$, "checkCancel", 
function () {
if (this.monitor != null && this.monitor.isCanceled ()) throw  new org.eclipse.core.runtime.OperationCanceledException ();
});
Clazz.defineMethod (c$, "checkCancelWithinCompiler", 
function () {
if (this.monitor != null && this.monitor.isCanceled () && !this.cancelling) {
this.setCancelling (true);
throw  new org.eclipse.jdt.internal.compiler.problem.AbortCompilation (true, null);
}});
Clazz.defineMethod (c$, "compiled", 
function (unit) {
var message = org.eclipse.jdt.internal.core.util.Messages.bind (org.eclipse.jdt.internal.core.util.Messages.build_compiling, unit.resource.getFullPath ().removeLastSegments (1).makeRelative ().toString ());
this.subTask (message);
this.updateProgressDelta (this.progressPerCompilationUnit);
this.checkCancelWithinCompiler ();
}, "net.sf.j2s.core.builder.SourceFile");
Clazz.defineMethod (c$, "done", 
function () {
($t$ = net.sf.j2s.core.builder.BuildNotifier.NewErrorCount = this.newErrorCount, net.sf.j2s.core.builder.BuildNotifier.prototype.NewErrorCount = net.sf.j2s.core.builder.BuildNotifier.NewErrorCount, $t$);
($t$ = net.sf.j2s.core.builder.BuildNotifier.FixedErrorCount = this.fixedErrorCount, net.sf.j2s.core.builder.BuildNotifier.prototype.FixedErrorCount = net.sf.j2s.core.builder.BuildNotifier.FixedErrorCount, $t$);
($t$ = net.sf.j2s.core.builder.BuildNotifier.NewWarningCount = this.newWarningCount, net.sf.j2s.core.builder.BuildNotifier.prototype.NewWarningCount = net.sf.j2s.core.builder.BuildNotifier.NewWarningCount, $t$);
($t$ = net.sf.j2s.core.builder.BuildNotifier.FixedWarningCount = this.fixedWarningCount, net.sf.j2s.core.builder.BuildNotifier.prototype.FixedWarningCount = net.sf.j2s.core.builder.BuildNotifier.FixedWarningCount, $t$);
this.updateProgress (1.0);
this.subTask (org.eclipse.jdt.internal.core.util.Messages.build_done);
if (this.monitor != null) this.monitor.done ();
this.previousSubtask = null;
if ("tizen".equals (this.j2sDeployMode)) {
if (net.sf.j2s.core.builder.BuildNotifier.NewErrorCount != 0) {
net.sf.j2s.core.utils.CorePluginConsole.error ("Build", "Project '%1$s' could not be built for '%2$s'", [this.currentProject.getName (), this.j2sDeployMode]);
org.eclipse.ui.PlatformUI.getWorkbench ().getDisplay ().asyncExec (((Clazz.isClassDefined ("net.sf.j2s.core.builder.BuildNotifier$1") ? 0 : net.sf.j2s.core.builder.BuildNotifier.$BuildNotifier$1$ ()), Clazz.innerTypeInstance (net.sf.j2s.core.builder.BuildNotifier$1, this, null)));
} else {
if ("Build".equals (this.builderMode)) {
net.sf.j2s.core.utils.CorePluginConsole.success ("Build", "Project '%1$s' has been built for '%2$s' successfully.", [this.currentProject.getName (), this.j2sDeployMode]);
}}} else if ("browser".equals (this.j2sDeployMode) && "Build".equals (this.builderMode)) {
if (net.sf.j2s.core.builder.BuildNotifier.NewErrorCount != 0) {
net.sf.j2s.core.utils.CorePluginConsole.error ("Build", "Project '%1$s' could not be built for '%2$s'", [this.currentProject.getName (), this.j2sDeployMode]);
} else {
net.sf.j2s.core.utils.CorePluginConsole.success ("Build", "Project '%1$s' has been built for '%2$s' successfully.", [this.currentProject.getName (), this.j2sDeployMode]);
}}try {
this.currentProject.refreshLocal (2, null);
} catch (e) {
if (Clazz.instanceOf (e, org.eclipse.core.runtime.CoreException)) {
e.printStackTrace ();
} else {
throw e;
}
}
});
Clazz.defineMethod (c$, "problemsMessage", 
function () {
var numNew = this.newErrorCount + this.newWarningCount;
var numFixed = this.fixedErrorCount + this.fixedWarningCount;
if (numNew == 0 && numFixed == 0) return "";
var displayBoth = numNew > 0 && numFixed > 0;
var buffer =  new StringBuffer ();
buffer.append ('(');
if (numNew > 0) {
buffer.append (org.eclipse.jdt.internal.core.util.Messages.build_foundHeader);
buffer.append (' ');
if (displayBoth || this.newErrorCount > 0) {
if (this.newErrorCount == 1) buffer.append (org.eclipse.jdt.internal.core.util.Messages.build_oneError);
 else buffer.append (org.eclipse.jdt.internal.core.util.Messages.bind (org.eclipse.jdt.internal.core.util.Messages.build_multipleErrors, String.valueOf (this.newErrorCount)));
if (displayBoth || this.newWarningCount > 0) buffer.append (" + ");
}if (displayBoth || this.newWarningCount > 0) {
if (this.newWarningCount == 1) buffer.append (org.eclipse.jdt.internal.core.util.Messages.build_oneWarning);
 else buffer.append (org.eclipse.jdt.internal.core.util.Messages.bind (org.eclipse.jdt.internal.core.util.Messages.build_multipleWarnings, String.valueOf (this.newWarningCount)));
}if (numFixed > 0) buffer.append (", ");
}if (numFixed > 0) {
buffer.append (org.eclipse.jdt.internal.core.util.Messages.build_fixedHeader);
buffer.append (' ');
if (displayBoth) {
buffer.append (String.valueOf (this.fixedErrorCount));
buffer.append (" + ");
buffer.append (String.valueOf (this.fixedWarningCount));
} else {
if (this.fixedErrorCount > 0) {
if (this.fixedErrorCount == 1) buffer.append (org.eclipse.jdt.internal.core.util.Messages.build_oneError);
 else buffer.append (org.eclipse.jdt.internal.core.util.Messages.bind (org.eclipse.jdt.internal.core.util.Messages.build_multipleErrors, String.valueOf (this.fixedErrorCount)));
if (this.fixedWarningCount > 0) buffer.append (" + ");
}if (this.fixedWarningCount > 0) {
if (this.fixedWarningCount == 1) buffer.append (org.eclipse.jdt.internal.core.util.Messages.build_oneWarning);
 else buffer.append (org.eclipse.jdt.internal.core.util.Messages.bind (org.eclipse.jdt.internal.core.util.Messages.build_multipleWarnings, String.valueOf (this.fixedWarningCount)));
}}}buffer.append (')');
return buffer.toString ();
});
Clazz.defineMethod (c$, "setCancelling", 
function (cancelling) {
this.cancelling = cancelling;
}, "~B");
Clazz.defineMethod (c$, "setProgressPerCompilationUnit", 
function (progress) {
this.progressPerCompilationUnit = progress;
}, "~N");
Clazz.defineMethod (c$, "subTask", 
function (message) {
var pm = this.problemsMessage ();
var msg = pm.length == 0 ? message : pm + " " + message;
if (msg.equals (this.previousSubtask)) return ;
if (this.monitor != null) this.monitor.subTask (msg);
this.previousSubtask = msg;
}, "~S");
Clazz.defineMethod (c$, "updateProblemCounts", 
function (newProblems) {
for (var i = 0, l = newProblems.length; i < l; i++) if (newProblems[i].isError ()) this.newErrorCount++;
 else this.newWarningCount++;

}, "~A");
Clazz.defineMethod (c$, "updateProblemCounts", 
function (oldProblems, newProblems) {
if (newProblems != null) {
next : for (var i = 0, l = newProblems.length; i < l; i++) {
var newProblem = newProblems[i];
if (newProblem.getID () == 536871362) continue ;var isError = newProblem.isError ();
var message = newProblem.getMessage ();
if (oldProblems != null) {
for (var j = 0, m = oldProblems.length; j < m; j++) {
var pb = oldProblems[j];
if (pb == null) continue ;var wasError = 2 == pb.getAttribute ("severity", 2);
if (isError == wasError && message.equals (pb.getAttribute ("message", ""))) {
oldProblems[j] = null;
continue next;}}
}if (isError) this.newErrorCount++;
 else this.newWarningCount++;
}
}if (oldProblems != null) {
next : for (var i = 0, l = oldProblems.length; i < l; i++) {
var oldProblem = oldProblems[i];
if (oldProblem == null) continue next;var wasError = 2 == oldProblem.getAttribute ("severity", 2);
var message = oldProblem.getAttribute ("message", "");
if (newProblems != null) {
for (var j = 0, m = newProblems.length; j < m; j++) {
var pb = newProblems[j];
if (pb.getID () == 536871362) continue ;if (wasError == pb.isError () && message.equals (pb.getMessage ())) continue next;}
}if (wasError) this.fixedErrorCount++;
 else this.fixedWarningCount++;
}
}}, "~A,~A");
Clazz.defineMethod (c$, "updateProgress", 
function (newPercentComplete) {
if (newPercentComplete > this.percentComplete) {
this.percentComplete = Math.min (newPercentComplete, 1.0);
var work = Math.round (this.percentComplete * this.totalWork);
if (work > this.workDone) {
if (this.monitor != null) this.monitor.worked (work - this.workDone);
this.workDone = work;
}}}, "~N");
Clazz.defineMethod (c$, "updateProgressDelta", 
function (percentWorked) {
this.updateProgress (this.percentComplete + percentWorked);
}, "~N");
Clazz.defineMethod (c$, "getBuilderMode", 
function () {
return this.builderMode;
});
Clazz.defineMethod (c$, "setBuilderMode", 
function (builderMode) {
this.builderMode = builderMode;
}, "~S");
c$.$BuildNotifier$1$ = function () {
Clazz.pu$h ();
c$ = Clazz.declareAnonymous (net.sf.j2s.core.builder, "BuildNotifier$1", null, Runnable);
Clazz.overrideMethod (c$, "run", 
function () {
var activeShell = org.eclipse.ui.PlatformUI.getWorkbench ().getActiveWorkbenchWindow ().getShell ();
var dialog =  new $wt.widgets.MessageBox (activeShell, 292);
dialog.setText ("Detect some compilation errors");
dialog.setMessage ("Do you really want to ignore these compile errors?");
DialogSync2Async.block (dialog, this, function () {
var returnCode = dialog.dialogReturn;
if (32 == returnCode) {
}});
return;
});
c$ = Clazz.p0p ();
};
Clazz.defineStatics (c$,
"J2S_DEPLOY_MODE_BROWSER", "browser",
"J2S_DEPLOY_MODE_TIZEN", "tizen",
"BUILD_TAG", "Build",
"CLEAN_TAG", "Clean",
"NewErrorCount", 0,
"FixedErrorCount", 0,
"NewWarningCount", 0,
"FixedWarningCount", 0);
});
